src/Autofac/Core/IComponentRegistration.cs

```diff
@@ -88,6 +88,14 @@ public interface IComponentRegistration : IDisposable
         /// </summary>
         bool IsAdapterForIndividualComponent { get; }
 
+        /// <summary>
+        /// Gets a value indicating whether this registration forms a 'Service Override', which forces itself as the default
+        /// registration for all service it provides.
+        /// Registrations with <see cref="IsServiceOverride"/> set to true will not be included when resolving a collection of a service
+        /// to get all implementations.
+        /// </summary>
+        bool IsServiceOverride { get; }
```

Comment:
Could the notion of being a composite root be a metadata item? (Would that be too slow on perf?)

Reply:
Or maybe a marker service `ICompositeRoot<T>` the way we have `IStartable`? That might make it easy to cache the list of composite roots available when a scope is built... might get tricky if someone adds a registration source that outputs those, though. It'd have to be internal. Hmmm.

Reply:
The only reason I didn't use a name like 'IsComposite' was because the concept of a service override could be used outside of composites; there's no reason a composite **has** to take an `IEnumerable` in, you could equally use it to fake things if you wanted, there's a few scenarios I can see in my head.

It's also not **just** a priority indicator, because we filter it and treat it differently, as you say. Putting it in metadata would definitely be slower.  I'm not a big fan of dumping important context info in Metadata if we can avoid it; metadata should really be just 'extra' stuff people want to add, rather than core stuff (I know that registration order is an exception to this...).

Other possible names for this property I can think of:

- IsVirtualRegistration - The registration is not a 'concrete' registration, so a 'virtual' , or 'non-concrete' one makes sense. On the other hand, it is a registration you can construct...
- IsFixedRegistration - The registration is 'fixed', in the sense that it's always at the top of the pile (sort of like fixed CSS positioning in my brain, where it's always there, but sits outside of layout).
- IsRedirectRegistration - Any requests for the service get 'redirected' to this registration, kind of like a URL redirect, but you can still get the list of registrations.

Reply:
I think one of the things I'm getting hung up on is that we explicitly don't _decorate_ these things. If I had a virtual registration, if it was just about ordering (fixed), or if it was simply redirecting, I'd still assume it's getting decorated. The only use case I can come up with where it _both_ needs ordering _and_ explicitly shouldn't be decorated is... composites.

What if it was more like an options flag? Instead of a simple Boolean, something like

```c#
[Flags]
public enum RegistrationOptions
{
  None = 0,
  OverrideOrFixedOrVirtualOrWhatever = 1,
  Composite = 2,
  SomeOtherThingWeAddLater = 4,
}
```

That might also address the growing parameter list - instead of adding more Booleans, we can add new flags.

Reply:
I like the `[Flags]` idea. Perhaps we can put `IsAdapterForIndividualComponent` in there as well, which will reduce the number of constructor parameters back down.

Reply:
Okay, so the flags changes are in; `RegistrationData` now has an `Options` property; this gets passed through and picked up on the `IComponentRegistration`.

It's much neater now; I've declared each 'property' of a registration that effects different behaviour as a separate flag, with a single combined flag for `Composite`:

```csharp
[Flags]
public enum RegistrationOptions
{
    /// <summary>
    /// No special options; default behaviour.
    /// </summary>
    None = 0,

    /// <summary>
    /// Indicates that this registration is 'fixed' as the default, ignoring all other registrations when determining the default registration for
    /// a service.
    /// </summary>
    Fixed = 2,

    /// <summary>
    /// Registrations with this flag will not be decorated.
    /// </summary>
    DisableDecoration = 4,

    /// <summary>
    /// Registrations with this flag will not be included in any collection resolves (i.e. <see cref="IEnumerable{TService}" /> and other collection types).
    /// </summary>
    ExcludeFromCollections = 8,

    /// <summary>
    /// Flag combination for composite registrations.
    /// </summary>
    Composite = Fixed | DisableDecoration | ExcludeFromCollections,
}
```

This should make it more flexible in the future.

Also, added benefit, I realised I can just remove IsAdapterForIndividualComponent entirely from Component Registrations; it was only used when looking up decorators, which we now do entirely differently in the pipelines world.